openedx/openedx-platform
# [FC-0009] New django app for copying and pasting OLX content in Studio

## Description

This implements https://github.com/openedx/modular-learning/issues/51 and https://github.com/openedx/modular-learning/issues/10

It is implemented as a more generic "staged content" area, where "staged content" is any content that isn't part of a particular course or library at the moment. Staged content is not editable; you have to paste it somewhere to edit it. We anticipate that "partial import" functionality may be another use case for staged content in addition to this clipboard functionality.

It's currently implemented as a django app plugin within edx-platform because it's highly coupled to platform internals and we want it tested as part of the test suite. However, if preferable we could move it to an external repo.

I am hoping to use #31903 to enforce that no other code imports the private API of this new app.

## Supporting information

See GitHub issue linked above.

### Analytics

When a user copies something to their clipboard, the CMS will log something like this:
```
2023-03-17 00:15:58,445 INFO 465 [openedx.core.djangoapps.content_staging.views] [user 4] [ip 172.18.0.1] views.py:142 - 
User 4 copied video component "block-v1:DemoX+D+EMO+type@video+block@56a60936c4c246ac861cfb4419cf235e" to their clipboard.
```

## TODOs:

- [x] Add some test cases
- [ ] Consolidate block_serializers to a single API - to do after, as https://github.com/openedx/modular-learning/issues/54
- [x] Use clipboard model and move source_context there.
- [x] Fix filename of downloaded OLX
- [x] Add logging for statistics/evaluation purposes

## Testing instructions

1. Go to http://studio.local.overhang.io:8001/admin/waffle/flag/ (or devstack/sandbox equivalent)
2. Add a new flag, `contentstore.enable_copy_paste_feature` and enable it for Everyone
3. Go to a Unit page in Studio and use the new actions menu (see #31853). The "Copy" action should now work.
4. You won't be able to paste, but you can go to http://studio.local.overhang.io:8001/api/content-staging/v1/clipboard/ to see the clipboard contents. That contains an `olx_url` that you can go to in turn to download the OLX as an .xml file.

### How to test asynchronous copying

![copying](https://user-images.githubusercontent.com/945577/225768633-035b4968-e3ce-4553-b549-e91cb05cd208.png)

Copying is currently done synchronously, but the backend and frontend code _do_ already support long-running asynchronous copy operations. To simulate this for testing:

<details>

<summary>Click to view instructions</summary>

First, make the following edit:

![Screenshot 2023-03-16 at 3 36 36 PM](https://user-images.githubusercontent.com/945577/225768357-711cdb51-4a26-4b19-96fa-2f3842253fc6.png)

Then copy an XBlock. The copying process will continue until you retrieve the OLX. To retrieve the OLX, open the Network inspector and in one of the recent "clipboard/" responses, you'll see the URL - go to the highlighted URL:

![Screenshot 2023-03-16 at 3 39 27 PM](https://user-images.githubusercontent.com/945577/225768617-1e44d9b2-b86c-4e1e-a032-071e0801d4d0.png)

You will see that the "Copying..." toast stays until you access that URL, which is how we signal the completion of our (faked) async operation.
</details>

## Deadline

None

## Other information

Private ref - MNG-3572
